### DISCUSSION TITLE
Zoom dependent level of detail

### DISCUSSION THREAD

**Discussion by gee4vee** [eclipse-glsp/glsp#594] [General] [answered]
I saw a slide about this in a presentation on YouTube about GLSP. Is this still being worked on and if so what is the status?

**Comment by planger** [comment, marked as answer]
Thanks for your interest. It is still being worked on, but it is not clear yet when and in which form it'll arrive in GLSP. It is the work of an external contributor.

However, please note that you can implement something like that "manually" already now on the client. Essentially you'd need to [query the zoom level](https://github.com/eclipse/sprotty/blob/18c4d4d5cbb801347b72625198c7dd4919bba2cb/packages/sprotty/src/features/viewport/zoom.ts#L41) in the view implementation on the client and then decide to e.g. create or omit creating SVG elements or invoke or not invoke `renderChildren()` accordingly to show more or less detail.

**Reply by N1k145** [reply]
@planger do you know if there is any progress on this?

**Reply by planger** [reply]
Yes and no! The work has concluded in the meantime, but a generic declarative approach to add the level of detail into the graph model didn't make it into the main branch (yet). This would need another iteration of design to clarify in more detail how to express these rules in the graph model without adding too much complexity.

This was also considered from the perspective that it actually can be implemented with custom code in the views quite easily and therefore it was not clear whether the value of a generic declarative approach outweighs the complexity it introduces. You can easily query the zoom level of the viewport in the view implementations and control whether a node or also parts of a view are to be rendered:

```typescript
import {
    getZoom,
    ...
} from '@eclipse-glsp/client'

@injectable()
export class YourView extends ShapeView {
  render(element: GNode, context: RenderingContext): VNode | undefined {
    if (getZoom(element) < 10) {
      return undefined;
    }
    ...
  }
}
```

**Reply by N1k145** [reply]
Thanks for the Feedback, at a first glance, your proposal works. But the prototype does a relay out when the hierarchy changes and is able to change the text size in this step, or at least it looks like this.
We will investigate this further in the coming weeks and see if we can recreate that behavior in our client.

I will report back if this works for us. If not, we are open to clean up the prototype and open a PR after discussions on the setup of the rules, if you are interested.

**Reply by planger** [reply]
Thanks, sure we certainly appreciate a PR (whether it is based on the prototype or not) that eases creating the semantic zoom functionality.
Regarding the font size: an easy way to implement that is to add a corresponding CSS class (e.g. small, medium, large) in the view implementation based on the zoom level and have CSS rules that based on the class assign a specific font size.

**Reply by david-pa** [reply]
@planger has there been anything officially added to GLSP as a solution for level-of-detail zoom? It's been a while, but I couldn't tell from reading over the docs if this had built-in support yet.

**Reply by planger** [reply]
@david-pa no unfortunately not. We've implemented a few specific semantic-zoom-like features in customer products already, but there hasn't been a good opportunity yet to extract a generic implementation into the core yet.

**Reply by david-pa** [reply]
@planger is there any way I can get the info on the contributor that originally started the work on the feature, or at least whoever has been working on it most recently? I'd like to be able to include that sort of conditional rendering in a project.

**Reply by planger** [reply]
The original paper in which we explored and prototypically implemented this approach is this one:

https://scholar.google.at/scholar_url?url=https://dl.acm.org/doi/pdf/10.1145/3550355.3552412&hl=de&sa=X&ei=Z07TZ-y7LdmlieoPir2teA&scisig=AFWwaebjfHG3TyKlXwZNY9KY84-d&oi=scholarr

But really the fundamental building block is what I sketched above, ie making the rendering conditional in the view implementation based on the zoom level. There are a few additional things to consider, like ensuring that the overall size of elements doesn't change to avoid shifts during zoom changes, and structuring the view code accordingly (eg NodeOverviewView, NodeDetailView, and a NodeSwitchView which delegates to one or the other based in the zoom level) to avoid mixing too many concerns in one view implementation.

I hope this helps!
